[MERGE AFTER SDK #159] Adding GCS dependency for backend and prompt service

harini-venkataraman opened the pull request:
## What
Adding dependency version needed for Backend and prompt service.
...

## Why
This will be needed while the integrator uses fsspec. Corresponding file storage dependency is to be added.
...

## How
Pinned the dependency in toml.
...

## Can this PR break any existing features. If yes, please list possible items. If no, please explain why. (PS: Admins do not merge the PR without this section filled)
No, these are additions of dependency
...

## Database Migrations
Not applicable.
...

## Env Config
Not applicable.
...

## Relevant Docs

-

## Related Issues or PRs
Not applicable.
...

## Dependencies Versions
Not applicable.
...

## Notes on Testing

In backend, this is already present as a transitive dependency through connectors.
Pinned the version.
...

## Screenshots

## Checklist

I have read and understood the [Contribution Guidelines](https://docs.unstract.com/unstract/contributing/unstract/).

ritwik-g commented on backend/pyproject.toml:
```diff
@@ -33,6 +33,7 @@ dependencies = [
     "social-auth-app-django==5.3.0", # For OAuth
     "social-auth-core==4.4.2", # For OAuth
     "unstract-sdk~=0.56.0rc4",
+    "gcsfs==2024.6.0",
```
@harini-venkataraman won't we need the dependencies for Azure and S3 if this is needed? @gaya3-zipstack @hari-kuriakose

ritwik-g replied:
@gaya3-zipstack I got a doubt. Shouldn't we be adding this in the SDK?

gaya3-zipstack replied:
@ritwik-g Adding it in SDK is a baggage for SDK as we are only  using fsspec APIs. The integrator can decide what filesystem is required and cann the dependencies accordingly.
Eg, tools need minio and not  GCS. Hence adding GCS in SDK will be unnecessary for tools.
However we could use a selective way to install SDK like explicitly mention what we need to be added. We could take it up later....

ritwik-g replied:
Okay but let's make sure to add the Azure and S3 dependencies as well.

harini-venkataraman replied:
s3 and azure is not used in backend yet, and we can add it when we make the integration.

ritwik-g replied:
@harini-venkataraman @gaya3-zipstack we will be requiring it as soon as we release it to production for on-prem customers. This is not a separate requirement. S3, GCS and Azure storage support for all 3 needs to be present. Testing wise we can test it later. So my suggestion would be to take care of this

ritwik-g replied:
Any way this is what I feel. If you think it is better to proceed without the same please go ahead.

harini-venkataraman replied:
https://github.com/Zipstack/unstract/pull/1106#issuecomment-2638452058
Adding the dependencies as optional groip from SDK.

harini-venkataraman replied:
I would suggest, we address azure, once integration is stable
CC @gaya3-zipstack

harini-venkataraman commented on prompt-service/pyproject.toml:
```diff
@@ -16,6 +16,8 @@ dependencies = [
     "python-dotenv==1.0.0",
     "unstract-sdk~=0.56.0rc4",
+    "gcsfs==2024.10.0",
+    "s3fs==2024.6.0",
```
Yes @gaya3-zipstack , gcfs is inline with he version in connectors and s3fs from tools.

harini-venkataraman replied:
As discussed, unpinned the versions.

ritwik-g commented on unstract/connectors/pyproject.toml:
```diff
@@ -14,12 +14,12 @@ dependencies = [
     "google-cloud-storage==2.9.0",
     # Filesystem connectors
-    "s3fs[boto3]==2024.6.0", # For Minio
+    "s3fs[boto3]", # For Minio
```
@gaya3-zipstack @harini-venkataraman I think we should pin it to a major version while allowing the minor and patch version selection. Not mentioning any version would not be a good idea.

cc: @hari-kuriakose @chandrasekharan-zipstack

ritwik-g commented on unstract/connectors/pyproject.toml:
```diff
@@ -14,12 +14,12 @@ dependencies = [
     "dropboxdrivefs==1.4.1", # For Dropbox
     "boxfs==0.2.1", # For Box
-    "gcsfs==2024.6.0", # For GoogleCloudStorage
+    "gcsfs", # For GoogleCloudStorage
```
@gaya3-zipstack @harini-venkataraman I think we should pin it to a major version while allowing the minor and patch version selection. Not mentioning any version would not be a good idea.

cc: @hari-kuriakose @chandrasekharan-zipstack

harini-venkataraman replied:
Addressed @ritwik-g

gaya3-zipstack commented on backend/prompt_studio/prompt_studio_core_v2/prompt_ide_base_tool.py:
```diff
@@ -20,11 +17,6 @@ def __init__(self, log_level: LogLevel = LogLevel.INFO, org_id: str = "") -> Non
         self.log_level = log_level
         self.org_id = org_id
         self.workflow_filestorage = None
-        if check_feature_flag_status(FeatureFlag.REMOTE_FILE_STORAGE):
```
@harini-venkataraman Wondering if we should initialise workflow_filestorage to GCS here. This will save us from doing checks like provider==gcs in the SDK PR.
This is based on the assumption that  PromptIdeBase tool is only used in the context of PromptStudio and no other tool
I think if we rename workflow_filestorage to a more generic name like filestorage, it would look much cleaner. But we can do that later.

harini-venkataraman replied:
Addressed

gaya3-zipstack commented on backend/pyproject.toml:
```diff
@@ -33,6 +33,7 @@ dependencies = [
     "social-auth-app-django==5.3.0", # For OAuth
     "social-auth-core==4.4.2", # For OAuth
     "unstract-sdk~=0.56.0rc4",
+    "gcsfs~=2024.10.0",
```
Unpinning here does not work is it?

ritwik-g replied:
@gaya3-zipstack please unpin at minor version level like below at every place

```suggestion
    "gcsfs~=2024.10",
```

ritwik-g replied:
@gaya3-zipstack also looking at this [commit](https://github.com/Zipstack/unstract/pull/1106/commits/9d5c60ba23b94dd54bedff340ab71741d886ec11) 

I think we should keep the min version to 6. For backend I think that is the only compatible version. So keep it 
 ```suggestion
    "gcsfs~=2024.6",
```

and try

ritwik-g commented on platform-service/pyproject.toml:
```diff
@@ -14,7 +14,7 @@ dependencies = [
     "requests>=2.31.0",
     "unstract-sdk~=0.56.0rc4",
-    "gcsfs==2024.10.0",
+    "gcsfs~=2024.10.0",
```
```suggestion
    "gcsfs~=2024.6",
```

ritwik-g commented on prompt-service/pyproject.toml:
```diff
@@ -16,6 +16,8 @@ dependencies = [
     "llama-index==0.12.8",
     "python-dotenv==1.0.0",
     "unstract-sdk~=0.56.0rc4",
+    "gcsfs~=2024.10.0",
```
```suggestion
    "gcsfs~=2024.6",
```

ritwik-g commented on unstract/connectors/pyproject.toml:
```diff
@@ -14,12 +14,12 @@ dependencies = [
     "dropboxdrivefs==1.4.1", # For Dropbox
     "boxfs==0.2.1", # For Box
-    "gcsfs==2024.6.0", # For GoogleCloudStorage
+    "gcsfs~=2024.10.0", # For GoogleCloudStorage
```
```suggestion
    "gcsfs~=2024.6", # For GoogleCloudStorage
```

hari-kuriakose reviewed (changes requested):
@harini-venkataraman If `gcsfs` and `s3fs` etc are required only for code in SDK, then it will not be good for the long term to include them as direct dependencies downstream.

However, keeping packages for all cloud providers as default dependencies in SDK will also lead to bloating.

Thus I suggest that we leverage optional dependencies in SDK `pyproject.toml` to have a section like the following:

```
[project.optional-dependencies]
aws = [
  "s3fs~=2024.10.0"
]
azure = [
]
gcp = [
  "gcsfs~=2024.6"
]
```

This will allow downstream services to install SDK selectively by having dependencies like `unstract-sdk`, `unstract-sdk[gcp]`, etc.

**NOTE**: It is relatively OK to release with only GCP support first.

harini-venkataraman commented:
> @harini-venkataraman If `gcsfs` and `s3fs` etc are required only for code in SDK, then it will not be good for the long term to include them as direct dependencies downstream.
> 
> However, keeping packages for all cloud providers as default dependencies in SDK will also lead to bloating.
> 
> Thus I suggest that we leverage optional dependencies in SDK `pyproject.toml` to have a section like the following:
> 
> ```
> [project.optional-dependencies]
> aws = [
>   "s3fs~=2024.10.0"
> ]
> azure = [
> ]
> gcp = [
>   "gcsfs~=2024.6"
> ]
> ```
> 
> This will allow downstream services to install SDK selectively by having dependencies like `unstract-sdk`, `unstract-sdk[gcp]`, etc.
> 
> **NOTE**: It is relatively OK to release with only GCP support first.
> 
> cc @ritwik-g @gaya3-zipstack @muhammad-ali-e

@hari-kuriakose Thanks for suggesting this. 
Have made the requested changes in https://github.com/Zipstack/unstract-sdk/pull/159

ritwik-g commented on unstract/connectors/pyproject.toml:
```diff
@@ -14,12 +14,12 @@ dependencies = [
     "dropboxdrivefs==1.4.1", # For Dropbox
     "boxfs==0.2.1", # For Box
-    "gcsfs==2024.6.0", # For GoogleCloudStorage
+    "gcsfs==2024.6", # For GoogleCloudStorage
```
```suggestion
    "gcsfs~=2024.6", # For GoogleCloudStorage
```

ritwik-g commented on unstract/connectors/pyproject.toml:
```diff
@@ -14,12 +14,12 @@ dependencies = [
     "google-cloud-storage==2.9.0",
     # Filesystem connectors
-    "s3fs[boto3]==2024.6.0", # For Minio
+    "s3fs[boto3]==2024.10.0", # For Minio
```
```suggestion
    "s3fs[boto3]~=2024.10.0", # For Minio
```

ritwik-g commented on tools/text_extractor/requirements.txt:
```diff
@@ -3,4 +3,4 @@
 unstract-sdk~=0.56.0rc4
 # Required for remote storage support
-s3fs[boto3]==2024.6.0
+s3fs[boto3]~=2024.10.0
```
```suggestion
s3fs[boto3]~=2024.10
```

ritwik-g commented on tools/structure/requirements.txt:
```diff
@@ -3,4 +3,4 @@
 unstract-sdk~=0.56.0rc4
 # Required for remote storage support
-s3fs[boto3]==2024.6.0
+s3fs[boto3]~=2024.10.0
```
```suggestion
s3fs[boto3]~=2024.10
```

ritwik-g commented on tools/classifier/requirements.txt:
```diff
@@ -3,4 +3,4 @@
 unstract-sdk~=0.56.0rc4
 # Required for remote storage support
-s3fs[boto3]==2024.6.0
+s3fs[boto3]~=2024.10.0
```
```suggestion
s3fs[boto3]~=2024.10
```

ritwik-g reviewed (commented):
I have added few suggestions. But include it or not I will leave it up to you folks

harini-venkataraman commented:
The dependencies have been moved to SDK as an optional dep - https://github.com/Zipstack/unstract-sdk/pull/159, and other changes are cherrypicked to https://github.com/Zipstack/unstract/pull/1124/
Addressed the review comments in the corresponding PRs, and support for Azure will be handled once the current versio  is stable. Closing the PR with this comment.
